File: CSharp/UnicodeFileToHtmlTextConverter/HtmlPagesConverter.cs
namespace TDDMicroExercises.UnicodeFileToHtmlTextConverter;

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

public class HtmlPagesConverter
{
    private string filename;
    private List<int> breaks = new List<int>();

    public HtmlPagesConverter(string filename)
    {
        this.filename = filename;

        this.breaks.Add(0);
        using (StreamReader reader = new StreamReader(this.filename))
        {
            int cumulativeCharCount = 0;
            string line;
            while ((line = reader.ReadLine()) != null)
            {
                cumulativeCharCount += line.Length + 1; // add one for the newline
                if (line.Contains("PAGE_BREAK"))
                {
                    int pageBreakPosition = cumulativeCharCount;
                    breaks.Add(pageBreakPosition);
                }
            }
        }
    }

    public string GetHtmlPage(int page)
    {
        using (StreamReader reader = new StreamReader(this.filename))
        {
            reader.BaseStream.Seek(breaks[page], SeekOrigin.Begin);
            StringBuilder htmlPage = new StringBuilder();
            string line;
            while ((line = reader.ReadLine()) != null)
            {
                if (line.Contains("PAGE_BREAK"))
                {
                    break;
                }

                htmlPage.Append(System.Web.HttpUtility.HtmlEncode(line));
                htmlPage.Append("<br />");
            }

            return htmlPage.ToString();
        }
    }

    public string GetFilename()
    {
        return this.filename;
    }
}

File: kotlin/src/main/kotlin/com/katas/HtmlPagesConverter.kt
package com.katas

import java.io.BufferedReader
import java.io.FileReader

class HtmlPagesConverter(
    val filename: String
) {
    private val breaks: MutableList<Int> = ArrayList()

    init {
        breaks.add(0)
        val reader = BufferedReader(FileReader(filename))
        var cumulativeCharCount = 0
        var line = reader.readLine()
        while (line != null) {
            cumulativeCharCount += line.length + 1 // add one for the newline
            if (line.contains("PAGE_BREAK")) {
                val page_break_position = cumulativeCharCount
                breaks.add(page_break_position)
            }
            line = reader.readLine()
        }
        reader.close()
    }

    fun getHtmlPage(page: Int): String {
        val reader = BufferedReader(FileReader(this.filename))
        reader.skip(breaks[page].toLong())
        val htmlPage = StringBuffer()
        var line = reader.readLine()
        while (line != null) {
            if (line.contains("PAGE_BREAK")) {
                break
            }
            htmlPage.append(StringEscapeUtils.escapeHtml(line))
            htmlPage.append("<br />")

            line = reader.readLine()
        }
        reader.close()
        return htmlPage.toString()
    }
}
